### PR TITLE
bpo-37271: Optimize bytecode multiple times until it cannot be optimized further

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue37271

### CODE REVIEW

**Comment by nedbat** [comment]
@pablogsal Any chance you want to help get #13600 finished?

**Comment by pablogsal** [comment]
@nedbat Sure, I will talk with Victor to see if we can push it forward.

**Comment by pablogsal** [comment]
@serhiy-storchaka What do you think about this approach?

File: Python/peephole.c [hunk 1]
```diff
@@ -536,3 +527,45 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
...
+                Py_DECREF(old_bytecode);
+            }
+            return NULL;
+        }
+        compare_bytecode = PyObject_RichCompareBool(old_bytecode, bytecode, Py_NE);
+        if (old_bytecode != code) {
+            Py_DECREF(old_bytecode);
+        }
+        if (compare_bytecode == -1) {
+            break;
```
Comment (ZackerySpytz) [inline Python/peephole.c]: I don't think the `PyObject_RichCompareBool()` call is being handled properly. A return value of `-1` means that an error occurred, so `PyCode_Optimize()` should return `NULL` in that case.

File: Python/peephole.c [hunk 2]
```diff
@@ -536,3 +527,45 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
...
+
+PyObject *
+PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
+                PyObject *lnotab_obj)
+{
+    PyObject* bytecode = NULL;
+    PyObject* old_bytecode = code;
+    int compare_bytecode;
+    do {
+        bytecode = old_bytecode;
```
Comment (ZackerySpytz) [inline Python/peephole.c]: This assignment is unneeded.

File: Python/peephole.c [hunk 3]
```diff
@@ -536,3 +527,45 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
...
+   single basic block.  All transformations keep the code size the same or
+   smaller.  For those that reduce size, the gaps are initially filled with
+   NOPs.  Later those NOPs are removed and the jump addresses retargeted in
+   a single pass. */
+
+PyObject *
+PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
+                PyObject *lnotab_obj)
+{
+    PyObject* bytecode = NULL;
```
Comment (ZackerySpytz) [inline Python/peephole.c]: The declaration of `bytecode` can be moved into the loop.

File: Python/peephole.c [hunk 4]
```diff
@@ -536,3 +527,45 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
...
+    PyObject* old_bytecode = code;
```
Comment (ZackerySpytz) [inline Python/peephole.c]:
```suggestion
    PyObject *old_bytecode = code;
```

File: Python/peephole.c [hunk 5]
```diff
@@ -536,3 +527,45 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
...
+   encoding for gaps >= 255.
+
+   Optimizations are restricted to simple transformations occurring within a
+   single basic block.  All transformations keep the code size the same or
+   smaller.  For those that reduce size, the gaps are initially filled with
+   NOPs.  Later those NOPs are removed and the jump addresses retargeted in
+   a single pass. */
+
+PyObject *
+PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
```
Comment (ZackerySpytz) [inline Python/peephole.c]:
```suggestion
PyCode_Optimize(PyObject *code, PyObject *consts, PyObject *names,
```

File: Python/peephole.c (hunk 3 above)
Comment (pablogsal) [reply Python/peephole.c]: I am returning bytecode outside the loop, so that would need to be placed inside as well and I prefer not to do it in case the function needs some cleanup in the future.

File: Python/peephole.c (hunk 2 above)
Comment (pablogsal) [reply Python/peephole.c]: Yup, but it needs a change in the initialization of `bytecode` as `optimize_bytecode_once` cannot receive `NULL`.

**Comment by pablogsal** [comment]
I have rebased and address the feedback. Also, I have added a max cap of iterations (although is not needed, it can be proven that the while loop will always finish) to make sure the time expended in the loop is always bounded.

File: Python/peephole.c [hunk 6]
```diff
@@ -21,6 +21,7 @@
         (ABSOLUTE_JUMP(_Py_OPCODE(arr[i])) ? 0 : i+1))
 #define ISBASICBLOCK(blocks, start, end) \
     (blocks[start]==blocks[end])
+#define MAX_BYTECODE_OPT_ITERS 5
```
Comment (markshannon) [inline Python/peephole.c]: This seems arbitrary, and quite low.
What the rationale for it being 5?

File: Python/peephole.c [hunk 7]
```diff
@@ -474,10 +475,14 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
         unsigned int offset_delta, new_offset;
         cum_orig_offset += lnotab[i];
         assert(cum_orig_offset % sizeof(_Py_CODEUNIT) == 0);
-        new_offset = blocks[cum_orig_offset / sizeof(_Py_CODEUNIT)] *
-                sizeof(_Py_CODEUNIT);
-        offset_delta = new_offset - last_offset;
-        assert(offset_delta <= 255);
+        index = cum_orig_offset / sizeof(_Py_CODEUNIT);
```
Comment (markshannon) [inline Python/peephole.c:485]: Why is this change necessary?

**Review by markshannon — COMMENTED** [review]
Looks good in general, just a couple of questions.

File: Python/peephole.c (hunk 6 above)
Comment (pablogsal) [reply Python/peephole.c]: This is the average change in the biggest files when compiling the CPython test suite and code and is a number lower enough so it does not take a time increment that is big enough to notice.

Technically you can leave it unbounded because the change will always make the bytecode smaller, but this will assure that is small enough to not be a difference.

File: Python/peephole.c (hunk 7 above)
Comment (pablogsal) [reply Python/peephole.c:485]: Because the current line processing produces temporary indexes that violate that the `offset_delta < 255` when you run the optimizer multiple times. You can comment this change and experiment for yourself. This is the smaller and less invasive change to the peephole optimizer itself that I can think of, but I welcome any idea around this,

File: Python/peephole.c (hunk 6 above)
Comment (markshannon) [reply Python/peephole.c]: Can you add a brief comment to the code to this effect.

File: Python/peephole.c (hunk 7 above)
Comment (markshannon) [reply Python/peephole.c:485]: Again, can you add a brief comment to the code. thanks

File: Python/peephole.c (hunk 6 above)
Comment (pablogsal) [reply Python/peephole.c]: Will do :)

File: Python/peephole.c (hunk 7 above)
Comment (pablogsal) [reply Python/peephole.c:485]: Will do :)

**Comment by pablogsal** [comment]
@markshannon I have added some comments as requested, check if they look good to you :)

**Comment by pablogsal** [comment]
CC: @Yhg1s

**Comment by pablogsal** [comment]
This is the code that is failing the assert currently:

```
def f(cond1, cond2):
    while 1:
        return 3
    while 1:
        return 5
    return 6
```

**Comment by Yhg1s** [comment]
GH-15970 should fix the lnotab issue.
